darshan-util/pydarshan/darshan/cli/summary.py

```diff
@@ -0,0 +1,235 @@
+            mod_version = self.report.modules[mod]["ver"]
+            # retrieve the buffer size converted to KiB
+            mod_buf_size = self.report.modules[mod]["len"] / 1024
+            # create the key/value pairs for the dictionary
+            key = f"{mod} (ver={mod_version})"
+            val = f"{mod_buf_size:.2f} KiB"
+            module_dict[key] = val
+
+        # convert the module dictionary into a dataframe
+        module_df = pd.DataFrame.from_dict(data=module_dict, orient="index")
+        # write out the table in html
+        self.module_table = module_df.to_html(header=False, border=0)
+
+    def get_stylesheet(self):
+        """
+        Retrieves the locally stored CSS.
+        """
+        # get the path to the style sheet
+        with importlib_resources.path(darshan.cli, "style.css") as path:
+            # collect the css entries
+            with open(path, "r") as f:
+                self.stylesheet = "".join(f.readlines())
+
+
+def setup_parser(parser: Any):
+    """
+    Configures the command line arguments.
+
+    Parameters
+    ----------
+    parser : command line argument parser.
+
+    """
+    parser.add_argument(
+        "log_path",
+        type=str,
+        help="Specify path to darshan log.",
+    )
+    parser.add_argument("--output", type=str, help="Specify output filename.")
+
+
+def main(args: Union[Any, None] = None):
+    """
+    Generates a Darshan Summary Report.
+
+    Parameters
+    ----------
+    args: command line arguments.
+
+    """
+    if args is None:
+        parser = argparse.ArgumentParser(description="")
+        setup_parser(parser)
+        args = parser.parse_args()
+
+    log_path = args.log_path
+
+    if args.output is None:
+        # if no output is provided, use the log file
+        # name to create the output filename
+        log_filename = os.path.splitext(os.path.basename(log_path))[0]
+        report_filename = f"{log_filename}_report.html"
+    else:
+        report_filename = args.output
+
+    # collect the report data to feed into the template
+    report_data = ReportData(log_path=log_path)
+
+    with importlib_resources.path(darshan.cli, "") as path:
+        # get the path to the base template
+        base_path = os.path.join(str(path), "base.html")
+        # load a template object using the base template
+        template = Template(filename=base_path)
+        # render the base template
+        stream = template.render(report_data=report_data)
+        with open(report_filename, "w") as f:
+            # print a message so users know where to look for their report
+            save_path = os.path.join(os.getcwd(), report_filename)
+            print(
+                f"Report generated successfully. \n"
+                f"Saving report at location: {save_path}"
+            )
+            # save the rendered html
+            f.write(stream)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
I would normally try to place `argparse` stuff in this block, then have `main()` accept the arguments that were pulled out of the original command line. Wouldn't be surprised at all if some complications arose because of indirection here.

That may also substantially simplify the task of making `main()` directly/elegantly testable with regression suite without resorting to direct calls of the script via command line ugliness (gh-489).

Reply:
Having main() not require any of the argument parsing seems to me as well!